doc/dualtor/ICMP_Hardware_Offload_and_Protecion.md

```diff
@@ -0,0 +1,289 @@
+
+ICMP Hardware offload overcomes this limitation using the NPU hardware in the ToR switches to transmit and receive the ICMP packets and monitor the ICMP sessions for link state. This hardware based monitoring reduces the link state detection time.
+
+Protection switching reduces the switching time by avoiding reprogramming of routes and using Next Hop Protection Group in NPU hardware to toggle the traffic destination.
+
+Following diagram shows the cluster topology of Dual-ToR architecture for reference. Both ToRs use same loopback IP address as source IP in the ICMP echo packets.
+<div align="center"> <img src=image/cluster_topology.png width=600 /> </div>
+
+### Requirement
+Current link state detection time with software based ICMP sessions is in the order of 200-400 milliseconds. Requirement for ICMP Hardware Offload is to bring this down to < 10 milliseconds.
+
+Currently switching time after link state change is detected, is around 25-50 milliseconds. Note this depends on route scale. Requirement for FRR protection switching using Nexthop Protection Group is to keep it consistent irrespective of the route scale.
+
+### High Level Component and Requirements
+#### LinkMgrd Requirements
+LinkMgrd is the central component that runs in MUX docker and is responsible for managing ICMP echo sessions and link state. Link prober sub-component in Linkmgrd is responsible for the ICMP echo sessions.
+* Requirements
+   * Add / Remove Hardware ICMP echo session in App DB based on mux cable and link prober config entries.
+   * Consume ICMP echo session state from State DB produced by Orchagent and update mux state.
+   * Create two hardware sessions per mux port / interface one for self and another one to monitor the peer replies.
+   * Generate and handle ICMP packets with TLVs in software.
+
+Following diagram shows the hardware ICMP echo sessions between the TORs and Server.
+<div align="center"> <img src=image/hw_icmp_echo_sessions.png width=600 /> </div>
+
+#### Orchagent Requirement
+Currently orchagent creates tunnel at initialization and add / removes routes to forward traffic to peer ToR via a IpinIP tunnel when linkmgrd switches state to standby / active.
+* Requirements
+   * Create / Remove hardware ICMP echo sessions by consuming entries in App DB.
+   * Consume ICMP echo session notification from SAI and update session state in State DB.
+   * Create Protection Next Hop group for FRR switchover based on config.
+   * Initiate traffic switchover by toggling members of Protection Next Hop group by consuming link state produced by LinkMgrd.
+
+#### SaiRedis Requirement
+A new SAI specification was added to support ICMP Echo offload sessions in SAI version 1.12.
+* Requirements
+  * Support all SAI attributes of sai icmp echo session from saiicmpecho.h.
+  * Support SAI switch notification for sai icmp echo sessions.
+
+## Detailed Design

```

Comment:
`ASIC_DB` should be the only intermediary between `IcmpOrch` and `syncd`, the link between `IcmpOrch` and `SAI-Redis` looks redundant here.